### PR TITLE
add soft delete with CBS and CVM

### PR DESCRIPTION
**Pull request by gailwang**
add soft delete with CBS and CVM

### CODE REVIEW

**Comment by gailwang** [comment]
```
CVM apply & update renew flag

terraform apply
2020/05/15 15:19:10 [WARN] Log levels other than TRACE are currently unreliable, and are supported only for backward compatibility.
  Use TF_LOG=TRACE to see Terraform's internal logs.
  ----
tencentcloud_key_pair.random_key: Refreshing state... [id=skey-iadwe0d7]
data.tencentcloud_availability_zones.my_favorate_zones: Refreshing state...
data.tencentcloud_images.my_favorate_image: Refreshing state...
data.tencentcloud_instance_types.my_favorate_instance_types: Refreshing state...

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # tencentcloud_instance.foo[0] will be created
  + resource "tencentcloud_instance" "foo" {
      + allocate_public_ip                      = false
      + availability_zone                       = "ap-guangzhou-3"
      + create_time                             = (known after apply)
      + disable_monitor_service                 = true
      + disable_security_service                = false
      + expired_time                            = (known after apply)
      + force_delete                            = false
      + id                                      = (known after apply)
      + image_id                                = "img-1u6l2i9l"
      + instance_charge_type                    = "PREPAID"
      + instance_charge_type_prepaid_period     = 1
      + instance_charge_type_prepaid_renew_flag = "NOTIFY_AND_AUTO_RENEW"
      + instance_name                           = "terraform-brickzzhang-testing"
      + instance_status                         = (known after apply)
      + instance_type                           = "S1.SMALL1"
      + internet_charge_type                    = "TRAFFIC_POSTPAID_BY_HOUR"
      + internet_max_bandwidth_out              = 2
      + key_name                                = "skey-iadwe0d7"
      + private_ip                              = (known after apply)
      + project_id                              = 0
      + public_ip                               = (known after apply)
      + running_flag                            = true
      + security_groups                         = (known after apply)
      + subnet_id                               = (known after apply)
      + system_disk_id                          = (known after apply)
      + system_disk_size                        = 50
      + system_disk_type                        = "CLOUD_PREMIUM"
      + vpc_id                                  = (known after apply)

      + data_disks {
          + data_disk_id         = (known after apply)
          + data_disk_size       = (known after apply)
          + data_disk_type       = (known after apply)
          + delete_with_instance = (known after apply)
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

tencentcloud_instance.foo[0]: Creating...
tencentcloud_instance.foo[0]: Still creating... [10s elapsed]
tencentcloud_instance.foo[0]: Creation complete after 11s [id=ins-cbtb9kxs]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.


terraform apply
2020/05/15 15:20:27 [WARN] Log levels other than TRACE are currently unreliable, and are supported only for backward compatibility.
  Use TF_LOG=TRACE to see Terraform's internal logs.
  ----
data.tencentcloud_availability_zones.my_favorate_zones: Refreshing state...
tencentcloud_key_pair.random_key: Refreshing state... [id=skey-iadwe0d7]
data.tencentcloud_images.my_favorate_image: Refreshing state...
data.tencentcloud_instance_types.my_favorate_instance_types: Refreshing state...
tencentcloud_instance.foo[0]: Refreshing state... [id=ins-cbtb9kxs]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # tencentcloud_instance.foo[0] will be updated in-place
  ~ resource "tencentcloud_instance" "foo" {
        allocate_public_ip                      = false
        availability_zone                       = "ap-guangzhou-3"
        create_time                             = "2020-05-15T07:19:26Z"
        disable_monitor_service                 = true
        disable_security_service                = false
        expired_time                            = "2020-06-15T07:19:26Z"
        force_delete                            = false
        id                                      = "ins-cbtb9kxs"
        image_id                                = "img-1u6l2i9l"
        instance_charge_type                    = "PREPAID"
        instance_charge_type_prepaid_period     = 1
      ~ instance_charge_type_prepaid_renew_flag = "NOTIFY_AND_AUTO_RENEW" -> "NOTIFY_AND_MANUAL_RENEW"
        instance_name                           = "terraform-brickzzhang-testing"
        instance_status                         = "RUNNING"
        instance_type                           = "S1.SMALL1"
        internet_charge_type                    = "TRAFFIC_POSTPAID_BY_HOUR"
        internet_max_bandwidth_out              = 2
        key_name                                = "skey-iadwe0d7"
        private_ip                              = "10.104.37.28"
        project_id                              = 0
        running_flag                            = true
        security_groups                         = [
            "sg-nltpbqg1",
        ]
        system_disk_id                          = "disk-2apf3c6i"
        system_disk_size                        = 50
        system_disk_type                        = "CLOUD_PREMIUM"
        tags                                    = {}
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

tencentcloud_instance.foo[0]: Modifying... [id=ins-cbtb9kxs]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 10s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 20s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 30s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 40s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 50s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 1m0s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 1m10s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 1m20s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 1m30s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 1m40s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 1m50s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 2m0s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 2m10s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 2m20s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 2m30s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 2m40s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 2m50s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 3m0s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 3m10s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 3m20s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 3m30s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 3m40s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 3m50s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 4m0s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 4m10s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 4m20s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 4m30s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 4m40s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 4m50s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 5m0s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 5m10s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 5m20s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 5m30s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 5m40s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 5m50s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 6m0s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 6m10s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 6m20s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 6m30s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 6m40s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 6m50s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 7m0s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 7m10s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 7m20s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 7m30s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 7m40s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 7m50s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 8m0s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 8m10s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 8m20s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 8m30s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 8m40s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 8m50s elapsed]
tencentcloud_instance.foo[0]: Still modifying... [id=ins-cbtb9kxs, 9m0s elapsed]
tencentcloud_instance.foo[0]: Modifications complete after 9m1s [id=ins-cbtb9kxs]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
```

**Comment by gailwang** [comment]
```
CBS apply & update &destroy with prepaid instance
terraform apply
2020/05/15 15:08:38 [WARN] Log levels other than TRACE are currently unreliable, and are supported only for backward compatibility.
  Use TF_LOG=TRACE to see Terraform's internal logs.
  ----

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # tencentcloud_cbs_storage.storage_upgrade will be created
  + resource "tencentcloud_cbs_storage" "storage_upgrade" {
      + attached           = (known after apply)
      + availability_zone  = "ap-guangzhou-3"
      + charge_type        = "POSTPAID_BY_HOUR"
      + force_delete       = false
      + id                 = (known after apply)
      + prepaid_period     = (known after apply)
      + prepaid_renew_flag = (known after apply)
      + project_id         = 0
      + snapshot_id        = (known after apply)
      + storage_name       = "tf-storage-upgrade"
      + storage_size       = 50
      + storage_status     = (known after apply)
      + storage_type       = "CLOUD_PREMIUM"
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

tencentcloud_cbs_storage.storage_upgrade: Creating...
tencentcloud_cbs_storage.storage_upgrade: Creation complete after 1s [id=disk-g2wgtrdw]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.

 terraform apply
2020/05/15 15:09:53 [WARN] Log levels other than TRACE are currently unreliable, and are supported only for backward compatibility.
  Use TF_LOG=TRACE to see Terraform's internal logs.
  ----
tencentcloud_cbs_storage.storage_upgrade: Refreshing state... [id=disk-g2wgtrdw]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # tencentcloud_cbs_storage.storage_upgrade will be updated in-place
  ~ resource "tencentcloud_cbs_storage" "storage_upgrade" {
        attached           = false
        availability_zone  = "ap-guangzhou-3"
      ~ charge_type        = "POSTPAID_BY_HOUR" -> "PREPAID"
        encrypt            = false
      ~ force_delete       = false -> true
        id                 = "disk-g2wgtrdw"
      + prepaid_period     = 1
      + prepaid_renew_flag = "NOTIFY_AND_MANUAL_RENEW"
        project_id         = 0
        storage_name       = "tf-storage-upgrade"
        storage_size       = 50
        storage_status     = "UNATTACHED"
        storage_type       = "CLOUD_PREMIUM"
        tags               = {}
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

tencentcloud_cbs_storage.storage_upgrade: Modifying... [id=disk-g2wgtrdw]
tencentcloud_cbs_storage.storage_upgrade: Modifications complete after 1s [id=disk-g2wgtrdw]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.

terraform destroy
2020/05/15 15:10:44 [WARN] Log levels other than TRACE are currently unreliable, and are supported only for backward compatibility.
  Use TF_LOG=TRACE to see Terraform's internal logs.
  ----
tencentcloud_cbs_storage.storage_upgrade: Refreshing state... [id=disk-g2wgtrdw]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  - destroy

Terraform will perform the following actions:

  # tencentcloud_cbs_storage.storage_upgrade will be destroyed
  - resource "tencentcloud_cbs_storage" "storage_upgrade" {
      - attached           = false -> null
      - availability_zone  = "ap-guangzhou-3" -> null
      - charge_type        = "PREPAID" -> null
      - encrypt            = false -> null
      - force_delete       = true -> null
      - id                 = "disk-g2wgtrdw" -> null
      - prepaid_period     = 1 -> null
      - prepaid_renew_flag = "NOTIFY_AND_MANUAL_RENEW" -> null
      - project_id         = 0 -> null
      - storage_name       = "tf-storage-upgrade" -> null
      - storage_size       = 50 -> null
      - storage_status     = "UNATTACHED" -> null
      - storage_type       = "CLOUD_PREMIUM" -> null
      - tags               = {} -> null
    }

Plan: 0 to add, 0 to change, 1 to destroy.

Do you really want to destroy all resources?
  Terraform will destroy all your managed infrastructure, as shown above.
  There is no undo. Only 'yes' will be accepted to confirm.

  Enter a value: yes

tencentcloud_cbs_storage.storage_upgrade: Destroying... [id=disk-g2wgtrdw]
tencentcloud_cbs_storage.storage_upgrade: Destruction complete after 3s

Destroy complete! Resources: 1 destroyed.

```

**Comment by gailwang** [comment]
```
CBS DataSource

 make testacc TESTARGS=" -run=TestAccTencentCloudCbsStoragesDataSource"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccTencentCloudCbsStoragesDataSource -timeout 120m
?       github.com/terraform-providers/terraform-provider-tencentcloud  [no test files]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-tencentcloud/gendoc   (cached) [no tests to run]
=== RUN   TestAccTencentCloudCbsStoragesDataSource
=== PAUSE TestAccTencentCloudCbsStoragesDataSource
=== CONT  TestAccTencentCloudCbsStoragesDataSource
--- PASS: TestAccTencentCloudCbsStoragesDataSource (2.74s)
PASS
ok      github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud     3.542s
?       github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/connectivity[no test files]
?       github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/internal/helper      [no test files]
?       github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/ratelimit   [no test files]

```

File: tencentcloud/resource_tc_cbs_storage.go [hunk 1]
```diff
@@ -237,6 +283,12 @@ func resourceTencentCloudCbsStorageRead(d *schema.ResourceData, meta interface{}
 	_ = d.Set("tags", flattenCbsTagsMapping(storage.Tags))
 	_ = d.Set("storage_status", storage.DiskState)
 	_ = d.Set("attached", storage.Attached)
+	_ = d.Set("charge_type", storage.DiskChargeType)
+	_ = d.Set("prepaid_renew_flag", storage.RenewFlag)
+	_ = d.Set("charge_type", storage.DiskChargeType)
```
Comment (Himer) [inline tencentcloud/resource_tc_cbs_storage.go:288]: 少了prepaid_period?charge_type重复

File: tencentcloud/service_tencentcloud_cbs.go [hunk 2]
```diff
@@ -419,6 +419,41 @@ func (me *CbsService) UnattachSnapshotPolicy(ctx context.Context, diskId, policy
...
 
+func (me *CbsService) ModifyDiskChargeType(ctx context.Context, storageId string, chargeType string, renewFlag string, period int) error {
+	logId := getLogId(ctx)
+	request := cbs.NewModifyDisksChargeTypeRequest()
+	request.DiskIds = []*string{&storageId}
+	request.DiskChargePrepaid = &cbs.DiskChargePrepaid{Period: helper.IntUint64(period), RenewFlag: &renewFlag}
+	ratelimit.Check(request.GetAction())
+	response, err := me.client.UseCbsClient().ModifyDisksChargeType(request)
+	if err != nil {
+		log.Printf("[CRITAL]%s api[%s] fail, request body [%s], reason[%s]\n",
```
Comment (oliverpei) [inline tencentcloud/service_tencentcloud_cbs.go:430]: 这些请求日志可以去掉，现在会打印sdk的请求日志

File: tencentcloud/service_tencentcloud_cbs.go [hunk 3]
```diff
@@ -419,6 +419,41 @@ func (me *CbsService) UnattachSnapshotPolicy(ctx context.Context, diskId, policy
...
+			logId, request.GetAction(), request.ToJsonString(), err.Error())
+		return err
+	}
+	log.Printf("[DEBUG]%s api[%s] success, request body [%s], response body [%s]\n",
```
Comment (oliverpei) [inline tencentcloud/service_tencentcloud_cbs.go:434]: 这些请求日志可以去掉，现在会打印sdk的请求日志

**Pull request merged**